File: Source/Remote Control Server/Command.vb
Public Class Command

    Public Const PRIORITY_LOW As Byte = 0 'UDP
    Public Const PRIORITY_MEDIUM As Byte = 1 'TCP
    Public Const PRIORITY_HIGH As Byte = 2 'TCP retry
    Public Const PRIORITY_INDISPENSABLE As Byte = 3 'TCP retry forever

    Public source As String = "Unknown"
    Public destination As String = "Unknown"
    Public priority As Byte = 0
    Public type As Byte
    Public data As Byte()
    Public api As Byte = Remote.latestApi

    Public Sub send()
        Try
            Network.sendCommand(Me)
        Catch ex As Exception
            Logger.add("Unable to send command to " & destination & vbNewLine & ex.ToString)
        End Try
    End Sub

    Public Sub process()
        Try
            If Authentication.isAuthenticated(source, Server.getApp(source).pin) Or isBroadcast() Then
                parse()
            Else
                Logger.add("Refused a command from " & source)
                Logger.add("Server protection is active")
            End If
        Catch ex As Exception
            Logger.add("Error while processing command: " & vbNewLine & ex.ToString)
        End Try
    End Sub

    Private Sub parse()
        'Detect used API version
        Try
            Dim commandByte As Byte = data(0)

            If commandByte.Equals(ApiV3.COMMAND_IDENTIFIER + 1) Then
                'Newer API than available
                api = Remote.latestApi
                'TODO: Notify user
                'TODO: Check for updates
            ElseIf commandByte.Equals(ApiV3.COMMAND_IDENTIFIER) Then
                'Android version 2.1.0 and up
                api = 3
            ElseIf commandByte.Equals(ApiV2.COMMAND_IDENTIFIER) Then
                'Android version 1.8.0 - 2.0.3
                api = 2
            Else
                'Older implementations
                api = 1
            End If

            'Logger.add("Detected API version: " & api)
            Network.commandCount += 1
            Remote.processCommand(Me)
        Catch ex As Exception
            Logger.add("Error while parsing command: " & vbNewLine & ex.ToString)
        End Try
    End Sub

    Public Function dataAsString() As String
        Return System.Text.Encoding.UTF8.GetString(data)
    End Function

    Public Sub log()
        Logger.add(source & ": " & dataAsString())
    End Sub

    Public Function isBroadcast() As Boolean
        If ApiV1.isBroadcast(Me) Or ApiV2.isBroadcast(Me) Or ApiV3.isBroadcast(Me) Then
            Return True
        Else
            Return False
        End If
    End Function

End Class

File: Source/MacOS/RemoteControlServer/command.cpp
#include "apiv1.h"
#include "apiv2.h"
#include "apiv3.h"
#include "authentication.h"
#include "command.h"
#include "logger.h"
#include "network.h"
#include "remote.h"
#include "server.h"

#include <QDebug>

Command::Command()
{
    source = "Unknown";
    destination = "Unknown";
    priority = 0;
    api = Remote::Instance()->latestApi;
    data = new QByteArray();
}

void Command::send()
{
    Network::Instance()->sendCommand(*this);
}

void Command::process()
{
    if (Authentication::Instance()->isAuthenticated(source, Server::Instance()->getApp(source)->pin) || isBroadcast() || isConnectionCommand())
    {
        parse();
    }
    else
    {
        Logger::Instance()->add("Refused a command from " + source);
        Logger::Instance()->add("Server protection is active");
    }
}

void Command::parse()
{
    if (data->length() > 0)
    {
        char commandByte = data->at(0);
        if (commandByte == ApiV3::Instance()->COMMAND_IDENTIFIER + 1)
        {
            api = Remote::Instance()->latestApi;
        }
        else if (commandByte == ApiV3::Instance()->COMMAND_IDENTIFIER)
        {
            api = 3;
        }
        else if (commandByte == ApiV2::Instance()->COMMAND_IDENTIFIER)
        {
            api = 2;
        }
        else
        {
            api = 1;
        }
        Network::Instance()->commandCount += 1;
        Remote::Instance()->processCommand(*this);
    }
    else
    {
        Logger::Instance()->add("Could not parse command");
    }
}

QString Command::dataAsString()
{
    return *data;
}

void Command::log()
{
    Logger::Instance()->add(source + ": " + dataAsString());
}

bool Command::isBroadcast()
{
    return (ApiV1::Instance()->isBroadcast(*this) || ApiV2::Instance()->isBroadcast(*this) || ApiV3::Instance()->isBroadcast(*this)) ? true : false;
}

bool Command::isConnectionCommand()
{
    return (ApiV3::Instance()->isConnectionCommand(*this)) ? true : false;
}
